cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java

```diff
@@ -711,29 +715,35 @@ private void addBranchPrograms(String node, WorkflowProgramAdder programAdder, b
   }
 
   private BatchPhaseSpec getPhaseSpec(String programName, PipelinePhase phase) {
+    Map<String, String> phaseConnectorDatasets = new HashMap<>();
     // if this phase uses connectors, add the local dataset for that connector if we haven't already
     for (StageSpec connectorInfo : phase.getStagesOfType(Constants.Connector.PLUGIN_TYPE)) {
       String connectorName = connectorInfo.getName();
-      String datasetName = connectorDatasets.get(connectorName);
-      if (datasetName == null) {
-        datasetName = "conn-" + connectorNum++;
+      if (!connectorDatasets.containsKey(connectorName)) {
+        String datasetName = "conn-" + connectorNum++;
         connectorDatasets.put(connectorName, datasetName);
+        phaseConnectorDatasets.put(connectorName, datasetName);
         // add the local dataset
         ConnectorSource connectorSource = new MultiConnectorSource(datasetName, null);
         connectorSource.configure(getConfigurer());
+      } else {
+        phaseConnectorDatasets.put(connectorName, connectorDatasets.get(connectorName));
```

Comment:
don't think this else clause is required.

Reply:
Same as before